consensus/enclave/impl/src/constant_time_token_map.rs

```diff
@@ -0,0 +1,363 @@
+    /// end of the buffer. This enables us to later discard some or all of them,
+    /// without revealing over side-channels which token id's we discarded.
+    ///
+    /// This function systematically moves all the key-value pairs with zero
+    /// value to the end, in constant time. The algorithm is like an
+    /// insertion sort without early escapes, which should be fine for small
+    /// buffers.
+    pub fn sort_zeroes_to_end(&mut self) {
+        if self.storage.len() < 2 {
+            return;
+        }
+
+        let zero = T::default();
+        for split_pos in (1..self.storage.len()).rev() {
+            let (left, right) = self.storage.split_at_mut(split_pos);
+            let next = &mut right[0];
+            for candidate in left.iter_mut() {
+                // Swap if the thing on the left is zero, and the right is nonzero
+                // (In an insertion sort, this is just, swap if left < right)
+                let should_swap = candidate.1.ct_eq(&zero) & !next.1.ct_eq(&zero);
+                ConditionallySelectable::conditional_swap(
+                    &mut next.0,
+                    &mut candidate.0,
+                    should_swap,
+                );
+                ConditionallySelectable::conditional_swap(
+                    &mut next.1,
+                    &mut candidate.1,
+                    should_swap,
+                );
+            }
+        }
+    }
+}
+
+impl<T: ConditionallySelectable + Default> AsRef<[(TokenId, T)]> for CtTokenMap<T> {
+    fn as_ref(&self) -> &[(TokenId, T)] {
+        self.storage.as_ref()
+    }
+}
+
+// Create a Ct token map from a sequence of key-value pairs.
+//
+// Note: If keys repeat in this sequence, it is not an error. The second
+// instance of the key will be silently controlling. This is similar to how rust
+// stdlib implements FromIter for maps.
+impl<T: ConditionallySelectable + Default> FromIterator<(TokenId, T)> for CtTokenMap<T> {
+    fn from_iter<I>(iter: I) -> Self
+    where
+        I: IntoIterator<Item = (TokenId, T)>,
+    {
+        Self {
+            storage: iter.into_iter().collect(),
+        }
+    }
+}
+
+// Create a Ct token map from a sequence of pairs of references to keys and
+// values.
+//
+// Note: If keys repeat in this sequence, it is not an error. The second
+// instance of the key will be silently controlling. This is similar to how rust
+// stdlib implements FromIter for maps.
+impl<'a, T: ConditionallySelectable + Default> FromIterator<(&'a TokenId, &'a T)>
+    for CtTokenMap<T>
+{
+    fn from_iter<I>(iter: I) -> Self
+    where
+        I: IntoIterator<Item = (&'a TokenId, &'a T)>,
+    {
+        Self {
+            storage: iter.into_iter().map(|(key, val)| (*key, *val)).collect(),
+        }
+    }
+}
+
+// Create a Ct token map from a sequence of keys. The values are all defaulted.
+//
+// Note: If keys repeat in this sequence, it is not an error, and the API will
+// behave as expected.
```

Comment:
I'm not sure what to expect here. Consider adding a test case exercising and documenting the behaviour of get, add, set, etc.